PlagiarismPlugin.inc.php

```diff
@@ -43,169 +58,292 @@ public function getDescription() {
+		$publication = $submission->getCurrentPublication(); /** @var Publication $publication */
+		$author = $publication->getPrimaryAuthor(); /** @var Author $author */
 
 		// try to get credentials for current context otherwise use default config
-        	$contextId = $context->getId();
-		list($username, $password) = $this->getForcedCredentials();
-		if (empty($username) || empty($password)) {
-			$username = $this->getSetting($contextId, 'ithenticateUser');
-			$password = $this->getSetting($contextId, 'ithenticatePass');
-		}
+		list($apiUrl, $apiKey) = $this->hasForcedCredentials()
+			? $this->getForcedCredentials()
+			: [
+				$this->getSetting($context->getId(), 'ithenticateApiUrl'), 
+				$this->getSetting($context->getId(), 'ithenticateApiKey')
+			];
+
+		import("plugins.generic.plagiarism.IThenticate");
 
-		$ithenticate = null;
-		try {
-			$ithenticate = new \bsobbe\ithenticate\Ithenticate($username, $password);
-		} catch (Exception $e) {
-			$this->sendErrorMessage($submission->getId(), $e->getMessage());
+		/** @var \IThenticate $ithenticate */
+		$ithenticate = new IThenticate(
+			$apiUrl,
+			$apiKey,
+			static::PLUGIN_INTEGRATION_NAME,
+			$this->getCurrentVersion()->getData('current')
+		);
+
+		// Try to validate and confirm the EULA on user's behalf
+		if (!$ithenticate->confirmEula($request->getUser(), $context)) {
+			$this->sendErrorMessage($submission->getId(), 'Could not confirm EULA at iThenticate.');
 			return false;
 		}
-		// Make sure there's a group list for this context, creating if necessary.
-		$groupList = $ithenticate->fetchGroupList();
-		$contextName = $context->getLocalizedName($context->getPrimaryLocale());
-		if (!($groupId = array_search($contextName, $groupList))) {
-			// No folder group found for the context; create one.
-			$groupId = $ithenticate->createGroup($contextName);
-			if (!$groupId) {
-				$this->sendErrorMessage($submission->getId(), 'Could not create folder group for context ' . $contextName . ' on iThenticate.');
-				return false;
-			}
+
+		// Create the submission at iThenticate's end
+		$submissionUuid = $ithenticate->submitSubmission($submission, $request->getUser(), $author);
+
+		if (!$submissionUuid) {
+			$this->sendErrorMessage($submission->getId(), 'Could not submit the submission at iThenticate.');
+			return false;
 		}
 
-		// Create a folder for this submission.
-		if (!($folderId = $ithenticate->createFolder(
-			'Submission_' . $submission->getId(),
-			'Submission_' . $submission->getId() . ': ' . $publication->getLocalizedTitle($publication->getData('locale')),
-			$groupId,
-			true,
-			true
-		))) {
-			$this->sendErrorMessage($submission->getId(), 'Could not create folder for submission ID ' . $submission->getId() . ' on iThenticate.');
+		// $submission->setData('ithenticate_id', $submissionUuid);
+		// $submissionDao->updateObject($submission);		
+		Services::get("submission")->edit($submission, [
+			'ithenticate_id' => $submissionUuid,
+		], $request);
```

Comment:
@asmecher not sure why the setting are not updated , can you take a look what I am missing ?

Reply:
Probably what's happening is that you're fetching and then storing the updated submission, then the form implementation that calls the hook is immediately clobbering the submission from its own in-memory object. Try using the in-memory object and letting the form persist your changes after you modify it:

```diff
diff --git a/PlagiarismPlugin.inc.php b/PlagiarismPlugin.inc.php
index 85d8b9f..68d831e 100644
--- a/PlagiarismPlugin.inc.php
+++ b/PlagiarismPlugin.inc.php
@@ -229,8 +228,9 @@ class PlagiarismPlugin extends GenericPlugin {
 		$request = Application::get()->getRequest(); /** @var Request $request */
 		$context = $request->getContext(); /** @var Context $context */
 		$siteDao = DAORegistry::getDAO('SiteDAO'); /** @var SiteDAO $siteDao */
-		$submissionDao = DAORegistry::getDAO('SubmissionDAO'); /** @var SubmissionDAO $submissionDao */
-		$submission = $submissionDao->getById($request->getUserVar('submissionId')); /** @var Submission $submission */
+
+		$form =& $args[0];
+		$submission = $form->submission;
 		$publication = $submission->getCurrentPublication(); /** @var Publication $publication */
 		$author = $publication->getPrimaryAuthor(); /** @var Author $author */
 
@@ -271,12 +271,7 @@ class PlagiarismPlugin extends GenericPlugin {
 			return false;
 		}
 
-		// $submission->setData('ithenticate_id', $submissionUuid);
-		// $submissionDao->updateObject($submission);		
-		import('classes.core.Services');
-		Services::get("submission")->edit($submission, [
-			'ithenticate_id' => $submissionUuid,
-		], $request);
+		$submission->setData('ithenticate_id', $submissionUuid);
 		
 		// Upload submission files for successfully created submission at iThenticate's end
 		if (!$ithenticate->uploadSubmissionFile($submissionUuid, $submission)) {
```

Reply:
@asmecher big thanks for the help on this one . Now all fixed .